### PR TITLE
[GIE Physical] Support LateProject Strategy when Fetching Properties.

### LINKED ISSUE #2460: [GIE Physical] Introduce LateProject strategy when fetching properties
**Is your feature request related to a problem? Please describe.**

On distributed storage, we may not able to fetch properties of vertices directly as it may locate on a remote server. Thus in GIE, we would pre-cache the properties in the vertices currently.

For example, in the query:

`g.V().out().as("a").out().out().limit(10).select("a").by("name")` 

In order to fetch "name" from "a", we would cache the "name" for "a" when `g.V().out().as("a")`, and carry the property values along with the query (`out().out().limit(10)`), until we use it in `select("a").by("name")`.  Obviously, it is not a good choice in such scenarios. 

Thus, we propose to introduce a new strategy for fetching properties, i.e., fetching it when used. We call it a `LateProject` strategy. Via this strategy,  we would fetch the properties right before `select("a").by("name")` in the above example.

### PR DESCRIPTION
**Pull request by BingqingLyu**
## What do these changes do?

As titled.

## Related issue number

Fixes #2460

### CODE REVIEW

File: interactive_engine/executor/ir/core/src/plan/physical.rs [hunk 1]
```diff
@@ -41,301 +41,331 @@ pub trait AsPhysical {
...
+                .into_iter()
+                .map(|column| column.into())
+                .collect(),
+            is_all_columns: columns_opt.is_all(),
+            limit: None,
+            predicate: None,
+            sample_ratio: 1.0,
+            extra: Default::default(),
+        };
+        let auxilia = pb::GetV { tag: tag_pb.clone(), opt: 4, params: Some(params), alias: tag_pb.clone() };
```
Comment (longbinlai) [inline interactive_engine/executor/ir/core/src/plan/physical.rs:61]: Better comment what is `opt: 4` .

File: interactive_engine/executor/ir/core/src/plan/physical.rs [hunk 2]
```diff
@@ -41,301 +41,331 @@ pub trait AsPhysical {
...
+        builder.get_v(auxilia);
     }
 }
 
-fn merge_query_params(merged_params: &mut pb::QueryParams, other_params: &mut pb::QueryParams) {
-    merge_query_param_columns(merged_params, other_params.is_all_columns, &other_params.columns);
-    other_params.columns.clear();
-    other_params.is_all_columns = false;
+// Fetch properties before used in select, order, dedup, group, join, and apply.
```
Comment (longbinlai) [inline interactive_engine/executor/ir/core/src/plan/physical.rs:66]: Can you explain more about how to  process late project. Use some examples?

File: interactive_engine/executor/ir/core/src/plan/physical.rs [hunk 3]
```diff
@@ -41,301 +41,331 @@ pub trait AsPhysical {
...
-) {
-    if merged_params.is_all_columns || is_all_columns {
-        merged_params.is_all_columns = true;
-    } else {
-        for column in columns {
-            if !merged_params.columns.contains(column) {
-                merged_params.columns.push(column.clone());
-            }
-        }
+fn process_tag_columns(builder: &mut JobBuilder, tag: Option<TagId>, columns_opt: ColumnsOpt) {
```
[the remark this answers is not in the document]
Comment (BingqingLyu) [reply interactive_engine/executor/ir/core/src/plan/physical.rs:44]: Physical plan.

File: interactive_engine/executor/ir/core/src/plan/physical.rs [hunk 4]
```diff
@@ -41,301 +41,331 @@ pub trait AsPhysical {
...
+    if columns_opt.len() > 0 {
+        let tag_pb = tag.map(|tag_id| (tag_id as i32).into());
+        builder.shuffle(tag_pb.clone());
```
[the remark this answers is not in the document]
Comment (BingqingLyu) [reply interactive_engine/executor/ir/core/src/plan/physical.rs:47]: This function would be called only when `plan_meta.is_partition()`

File: interactive_engine/executor/ir/core/src/plan/physical.rs [hunk 5]
```diff
@@ -41,301 +41,331 @@ pub trait AsPhysical {
...
     }
 }
 
 impl AsPhysical for pb::Select {
     fn add_job_builder(&self, builder: &mut JobBuilder, plan_meta: &mut PlanMeta) -> IrResult<()> {
+        // This is the case when g.V().out().has(xxx), which was like Source + EdgeExpand(ExpandV) + Filter in logical plan.
+        // This would be refined as:
+        // In Logical Plan: `Source + EdgeExpand(ExpandE) + GetV`
+        // In Physical Plan:
+        //       1. on distributed graph store, `Source + EdgeExpand(ExpandV) + Shuffle + Auxilia`
```
[the remark this answers is not in the document]
Comment (BingqingLyu) [reply interactive_engine/executor/ir/core/src/plan/physical.rs]: Updated as `GetV`

File: interactive_engine/executor/ir/core/src/plan/physical.rs [hunk 6]
```diff
@@ -41,301 +41,331 @@ pub trait AsPhysical {
...
-    other_params.columns.clear();
-    other_params.is_all_columns = false;
+// Fetch properties before used in select, order, dedup, group, join, and apply.
+fn post_process_vars(
+    builder: &mut JobBuilder, plan_meta: &mut PlanMeta, is_order_or_group: bool,
+) -> IrResult<()> {
+    let node_meta = plan_meta.get_curr_node_meta().unwrap();
+    let tag_columns = node_meta.get_tag_columns();
+    let len = tag_columns.len();
+    if len == 0 {
```
[the remark this answers is not in the document]
Comment (BingqingLyu) [reply interactive_engine/executor/ir/core/src/plan/physical.rs]: Done

File: interactive_engine/compiler/src/main/java/com/alibaba/graphscope/gremlin/integration/graph/RemoteTestGraph.java [hunk 7]
```diff
@@ -1466,6 +1466,22 @@
                 "g_V_matchXa_0sungBy_b__a_0writtenBy_c__b_writtenBy_d__c_sungBy_d__d_hasXname_GarciaXX",
         test = "org.apache.tinkerpop.gremlin.process.traversal.step.map.MatchTest",
         reason = "unsupported")
+@Graph.OptOut(
+        method = "g_V_order_byXnameX_name",
+        test = "org.apache.tinkerpop.gremlin.process.traversal.step.map.OrderTest",
+        reason = "unsupoorted")
```
[the remark this answers is not in the document]
Comment (BingqingLyu) [reply interactive_engine/compiler/src/main/java/com/alibaba/graphscope/gremlin/integration/graph/RemoteTestGraph.java]: Updated as "project (with shuffle) after order occurs bug"

File: interactive_engine/executor/ir/core/src/plan/physical.rs (hunk 1 above)
Comment (BingqingLyu) [reply interactive_engine/executor/ir/core/src/plan/physical.rs:62]: Done

File: interactive_engine/executor/ir/core/src/plan/physical.rs (hunk 3 above)
[the remark this answers is not in the document]
Comment (longbinlai) [reply interactive_engine/executor/ir/core/src/plan/physical.rs:44]: Then probably PhysicalBuilder is better than JobBuilder.

File: interactive_engine/executor/ir/core/src/plan/physical.rs (hunk 4 above)
[the remark this answers is not in the document]
Comment (longbinlai) [reply interactive_engine/executor/ir/core/src/plan/physical.rs:47]: Then why not put `plan_meta.is_partition()` in this function? 
Let the caller be free of the context. 
Moreover, better comment this function to improve the understanding.

File: interactive_engine/compiler/src/main/java/com/alibaba/graphscope/gremlin/integration/graph/RemoteTestGraph.java (hunk 7 above)
[the remark this answers is not in the document]
Comment (longbinlai) [reply interactive_engine/compiler/src/main/java/com/alibaba/graphscope/gremlin/integration/graph/RemoteTestGraph.java]: "Projection may introduce additional shuffling that can break the order. "

File: interactive_engine/executor/ir/core/src/plan/physical.rs [hunk 8]
```diff
@@ -41,301 +41,310 @@ pub trait AsPhysical {
...
     }
 }
 
-fn merge_query_params(merged_params: &mut pb::QueryParams, other_params: &mut pb::QueryParams) {
-    merge_query_param_columns(merged_params, other_params.is_all_columns, &other_params.columns);
-    other_params.columns.clear();
-    other_params.is_all_columns = false;
+// Fetch properties before used in Project, Select, Order, Dedup, Group, Join, and Apply.
+// e.g.,
+// Case 1: For singe property (except used in Order or GroupVal), e.g., g.V().out().as("a").out().out().out().select("a").by("name"),
```
Comment (longbinlai) [inline interactive_engine/executor/ir/core/src/plan/physical.rs:69]: `single` property. Can use chatgpt to help check typo fyi.

File: interactive_engine/executor/ir/core/src/plan/physical.rs (hunk 4 above)
[the remark this answers is not in the document]
Comment (BingqingLyu) [reply interactive_engine/executor/ir/core/src/plan/physical.rs]: Done. Move this into `post_process_vars()` function.

**Review by longbinlai — APPROVED** [review]

**Pull request merged**